parlai/core/torch_generator_agent.py

```diff
@@ -1655,15 +1772,22 @@ def __init__(self, k, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.k = k
 
-    def select_paths(self, logprobs, prior_scores, current_length):
+    def select_paths(self, logprobs, prior_scores, current_length) -> _PathSelection:
         values, indices = logprobs.topk(self.k, dim=-1)
         probs = torch.softmax(values, dim=-1)
         choices = torch.multinomial(probs, 1)[:, 0]
         hyp_ids = torch.arange(logprobs.size(0)).to(logprobs.device)
         tok_ids = indices[hyp_ids, choices]
         scores = values[hyp_ids, choices]
         best_scores = prior_scores.expand_as(scores) + scores
-        return (hyp_ids, tok_ids, best_scores)
+
+        return _PathSelection(
+            hypothesis_ids=hyp_ids,
+            token_ids=tok_ids,
+            scores=best_scores,
+            token_scores=scores.view(-1),
+            token_ranks=choices.view(-1) + 1,
```

Comment:
(pointers for review):

Observe that `topk` returns elements in sorted order so the (single) choice has a rank equal to itself.

Reply:
thanks for the pointer!